Recreate per-lane bams on demand (and delete them when done)

gsanders opened the pull request:
This PR is to begin the discussion on the deletion and recreation of per-lane bams as appropriate. It's quite possible we are not locking enough and race conditions exist. I'm also sure we could use some code cleanup. We could use a lot of help with both of these things!

We are creating a 'real world one-off' integration test, which will a model with two builds. In the first build we will have two instrument data. In the third build we will delete the bams for the first two instrument data and archive the merged bam. The second build should unarchive the merged bam, recreate the two per-lane bams, align the third instrument data, and merge it all. If this works we'll have some confidence that we are doing what is required, but I'd still want to give a long look into locking and race condition issues.

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1629,9 +1629,112 @@ sub verify_alignment_data {
+sub recreated_alignment_bam_file_paths {
+    my $self = shift;
+
+    my @bams = glob($self->output_dir . "/*.bam");
```
This could call `alignment_bam_file_paths` instead of repeating the implementation.

gsanders replied:
Agreed! Fixed!

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1629,9 +1629,112 @@ sub verify_alignment_data {
+    }
+
+    my $recreated_bam = File::Spec->join($self->output_dir, 'all_sequences.bam');
+    my $merged_bam    = $self->get_merged_bam_to_revivify_per_lane_bam;
```
I like `revivify` as a word (It's almost as if we're alignment necromancers!)  But it's weird that it's only used here when `recreate` is used everywhere else.

gsanders replied:
Yeah, we werent very happy with the method name `recreated_alignment_bam_file_paths` in general. I want the method name to be something that means 'I'll get you the file paths, whether they exist yet or not' as opposed to the method above that only returns them if they exist.

I agree that revivify is both cooler and perhaps more appropriate than recreate (arguably). Unsure if it's unnecessarily confusing / fancy. I agree the same word should be used everywhere and I'll fix that once we figure out a good name for this method in general.

tmooney replied:
If `revivify` is too fancy, I'd vote for `reconstitute`.

susannasiebert commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1629,9 +1629,112 @@ sub verify_alignment_data {
 
 sub alignment_bam_file_paths {
+    return glob(shift->output_dir . "/*.bam");
+}
```
I would use `File::Spec->join(shift->output_dir, "/*.bam")` here.

gsanders replied:
Good point. Fixed!

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1629,9 +1629,112 @@ sub verify_alignment_data {
+    else {
+        return grep {! defined $_->test_name}@merged;
+    }
+}
```
This could possibly be simplified to:
```perl
sub get_merged_alignment_results {
  my $self = shift;
  return Genome::InstrumentData::AlignmentResult::Merged->get(
    'inputs.value_id' => $self->instrument_data_id,
    test_name => $self->test_name,
  );
}
```

We do have to watch out for the "bad" query, but joining once each to inputs and params should be fine.

tmooney replied:
On the other hand... we still need to check if the result we found matches our other inputs and parameters!

gsanders replied:
Good point. I'm really glad you caught this. This seems like a larger discussion!

What I'd like to do is just check that all of the is_param and is_input properties match. However, we are looking at `collect_individual_alignments` in `Merged.pm`, and I see that the logic for going from Merged -> per-lane seems more complex. It'd be REALLY awesome to somehow leverage this method into doing the reverse for us, but that seems complex.

An inefficient and silly (but potentially lines of code efficient?) way would be: get all of my merged alignment results that match by instrument data id (as above), and then call `collect_individual_alignments` on each, and do a set intersection to find the merged results that I belong to, according to the merged results themselves, and return those.

One issue I see with doing this is that on line `349` of `Merged.pm`, for test name we use the environment variable or undef. I'd think we'd want to match the test names on the objects?

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1629,9 +1629,112 @@ sub verify_alignment_data {
+        die $self->error_message('Failed to get valid merged bam to recreate per lane bam.');
+    }
+
+    my $cmd = Genome::InstrumentData::AlignmentResult::Command::RecreatePerLaneBam->create(
```
Before we can safely reconstitute the BAM, we must reallocate enough space.  (This might result in the allocation being moved to a new volume.)

tmooney commented:
The most common scenario we need to watch out for is:

1. Some instrument data comes through and a build is triggered.
2. More instrument data comes through and a new build is triggered.
3. Both builds are within merge at the same time.
4. The first completes and removes the per-lane BAM.
5. The second build suddenly finds the per-lane BAM has vanished.

The two builds might also contend over which gets to write the `.header` and `.flagstat` files in the per-lane BAM's allocation.

To do this safely requires some sort of locking around the per-lane alignment results.

gsanders commented:
Added code to:
- Address review concerns about refactoring `get_merged_alignment_results`
- Address review concerns about resizing the disk allocation before and after revivifying the per-lane bam
- Filter out non-database objects when looking for merged alignment results. This prevents us from getting our unfinished merged result in the current process.

Tests pass but we get some warnings about allocation moving in `AlignmentResult_regenerate.t`. Unsure if we should be concerned about this since it's just a unit test but we are toying with fixes.

gsanders commented:
Current TODO:
1) Address the point about matching the per-lane result parameters to the merged results we're getting in `get_merged_alignment_results`
2) Address the point about locking

gsanders commented:
Added some more commits here. Corrected a few errors, did some misc cleanup.

The biggest "feature" that has been added since our last post here is that per-lane alignment results will now consider parameter matching when searching for a merged alignment result with which to recreate itself.

gsanders commented:
Fixed the failing test. I believe this uncovered a bug with the test. The test_name was getting set on the merged alignment results but not the per-lane alignment results. This would happen to work, because collect_individual_alignments did not care that the test names were matching between the merged AR and the per-lane AR. I believe this to be a mistake but I would hope code reviewers would confirm this for us!

gsanders commented:
Added a commit that causes per-lane bams to be recreated in temporary allocations which are deleted once done.

TODO:
- Refactor eval block into a Try::Tiny block, and clean up temp allocations in a 'finally' block.

- Perform locking to prevent anyone from trying to recreate per-lane bams before they have been merged the first time, since this would be impossible with no merged bam.

dufeiyu commented:
CI, test this

gsanders commented:
Feiyu fixed the failing test. We also added locking that we believe is sufficient.

We plan on running some more real world build tests around this to create some contention. The current plan is:

1) Build 1 has instrument data A and B
2) Build 2 has instrument data B and C
3) Start both builds at once
4) One of the two builds will obtain the lock on B, and perform the first merged result
5) The second build will wait until the first build is done merging and releases the lock on B. It will then obtain the lock on B, recreate the per-lane B bam from its merged result, and then merge B and C.

dufeiyu commented:
CI, test this

gsanders commented:
Pushed some more commits, might be nearing completion here. Running real world tests to see what happens.

susannasiebert commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1628,10 +1643,144 @@ sub verify_alignment_data {
+    my $smallest = $results[0];
+
+    for my $result (@results) {
+        my @current_id = $result->instrument_data;
```
Can we rename these two to `@current_instrument_data` and `@smallest_instrument_data`? Calling it id is confusing to me.

gsanders replied:
This has been done and pushed. Please let me know if you like the new way better or if it still needs work! Thank you.

susannasiebert commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -219,24 +226,70 @@ sub create {
+            }
+            $alignment->output_dir($alignment->_disk_allocation->absolute_path); #update if was moved
         }
     }
```
I would pull this whole for-loop into its own subroutine. And maybe have separate subroutines for creating the header file, creating the flagstat file, and the deletion.

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -219,24 +226,70 @@ sub create {
+                    die $self->error_message("Fail to run flagstat on $bam_path");
+                }
+            }
+            $self->debug_message("Now removing the per lane bam");
```
Should this go in a post-commit hook to limit the chances that the files get deleted but something goes wrong and the merged result doesn't actually get committed?

dufeiyu replied:
The transaction of merged alignment result does commit BEFORE this removal of per lane bam. It is in finally block above.

tmooney replied:
Although the transaction does commit, the merged result is not saved to the database until the "process" context commits.  In the meantime, it might be bad to delete the per-lane BAM.

gsanders replied:
Glad you pointed this out! I did not understand that previously. We will look into it. Sorry we doubted you, Allison. :)

gsanders replied:
We just pushed a commit to address this. Please take a look and let us know what you all think!

susannasiebert commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -707,4 +786,50 @@ sub scalar_property_from_underlying_alignment_results {
+        unless ($alignment->get_merged_alignment_results) {
+            my @bams = $alignment->alignment_bam_file_paths;
+            unless (@bams) {
+               die $self->error_message("Per lane alignment $id has neither merged results nor bam paths. Please create an apipe-support ticket for this.");
```
Maybe give some more detail as to what needs to be done, i.e., that the instrument data needs to be realigned.

gsanders replied:
Thanks! I pushed a commit to address this. Let me know what you think.

gsanders commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -707,4 +786,50 @@ sub scalar_property_from_underlying_alignment_results {
+
+            unless ($ENV{UR_DBI_NO_COMMIT}) {
+                my $lock_var = 'genome_instrument_data_alignment_result-merged-per-lane-'.$id.'/lock';
+                my $lock = Genome::Sys->lock_resource(
```
As discussed in the code review today, this could cause a deadlock when some number of builds are in contention for 2+ shared instrument data. If every lock had information about other locks, they could release/retry in a random amount of time if a deadlock is detected. Or there could be a tiebreaker where the lower numbered instrument data in a sort is allowed to have the lock first, and other builds release and retry in an hour. Something similar.

This potentially should be part of the locking system?

gsanders commented:
Accidentally rolled back a change because of git problems. Fixed in the latest commit 5e9b352.

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -514,19 +521,26 @@ sub create {
 
     # STEP 14: RESIZE THE DISK
     # TODO: move this into the actual original allocation so we don't need to do this
```
Do you still agree with this TODO?  Is it worth doing, or can the TODO be removed?

dufeiyu replied:
Removed this obsolete comment.

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1628,10 +1643,144 @@ sub verify_alignment_data {
+    my $merged_bam    = $self->get_merged_bam_to_revivify_per_lane_bam;
+
+    unless ($merged_bam and -s $merged_bam) {
+        die $self->error_message('Failed to get valid merged bam to recreate per lane bam.');
```
It will probably be helpful in the error message to know the id of the bam we were trying to recreate

dufeiyu replied:
Added

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1628,10 +1643,144 @@ sub verify_alignment_data {
+    );
+
+    unless ($cmd->execute) {
+        die $self->error_message('Failed to execute RecreatePerLaneBam');
```
It will probably be helpful in the error message to know the id of the bam we were trying to recreate, if that is not obvious elsewhere in the log

dufeiyu replied:
Added

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1628,10 +1643,144 @@ sub verify_alignment_data {
+
+# This method will recreate the per-lane bam file and return the path.
+# This must be provided an allocation into which the bam will go (so that it can be cleaned up in the parent process once it is done).
+# The calling process is responsible for cleaning up the allocation after we are done with it.
```
Why is this done in an allocation rather than in /tmp?

dufeiyu replied:
Sometimes /tmp doesn't have enough space.

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1628,10 +1643,144 @@ sub verify_alignment_data {
+    my $merged_result = $self->get_smallest_merged_alignment_result($self->get_unarchived_merged_alignment_results);
+
+    unless ($merged_result) {
+        $merged_result = $self->get_smallest_merged_alignment_result($self->get_merged_alignment_results);
```
Is get_merged_alignment_results guaranteed to return at least one thing or die?  If not, we'll need to check here that merged_result actually exists before calling _auto_unarchive.

dufeiyu replied:
Good point, added

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1628,10 +1643,144 @@ sub verify_alignment_data {
+    for my $merged_result (@merged_results) {
+        my @individual_results = $merged_result->collect_individual_alignments;
+        if (@individual_results) {
+            push @matching_results, $merged_result if grep{refaddr($_) == refaddr($self)}@individual_results;
```
What is the advantage of using refaddr over, say, id?

dufeiyu replied:
We meant to compare object vs object firstly. But id comparison should be good enough here. Changed

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult.pm:
```diff
@@ -1628,10 +1643,144 @@ sub verify_alignment_data {
+        test_name => $self->test_name,
+    );
+    my @filtered_results = $self->filter_non_database_objects(@results);
+    return $self->filter_non_matching_results(@filtered_results);
```
You explained this during the meeting but I forget why you need to do this reverse lookup.

gsanders replied:
The reasons why we do a reverse_lookup:
- Merged.pm already knows how to get its alignment results, but the logic is difficult to reverse and generalize and considers both test_name and parameters, as well as instrument data id
- Getting every merge that shares an instrument data id with me, and then letting Merge.pm do the reverse lookup solves this

So I think the 'pro' is that we don't have to invert and duplicate code. The 'con' is that it's confusing and inefficient (but probably runs fast enough that we do not care). I'm not convinced this is the very best solution, it's just what we could come up with so far. I think the best solution I can think of is to have a stable SoftwareResult::User relationship between per-lane and merged alignment results.

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -191,7 +196,9 @@ sub create {
             #just collect the BAMs for a merge
             for my $alignment (@alignments) {
-                push @bams_for_final_merge, $alignment->alignment_bam_file_paths;
+                my $temp_allocation = $self->_get_temp_allocation($alignment);
```
This line is duplicated from line 178  - should it be moved outside the if/else block?

gsanders replied:
I want to pull it out... but in our first attempt to do so we added more lines of code than we took out :p. So I'm not sure if it can be made more clean. I will look again but let me know if you have an idea. We may leave it as-is.

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -219,24 +226,70 @@ sub create {
+            $allocation->delete;
+        }
+        unless ($_) {
+            $tx->commit();
```
Does this mean the transaction won't be committed if the allocation deletes fail?  Is that what we want?  Or would it make sense to move the commit to the end of the try block?

gsanders replied:
It looks like based upon testing Try::Tiny in a script, you are correct. The commit will not run if something 'dies' in the delete.

It's possible to add multiple 'finally' blocks, and the LAST 'finally' to execute as part of this block needs a semicolon after its end brace.

Example for our own information:

```
my $var;
try {
    print "in the try\n";
    die "died in try";
    print "still printing in the TRY after death\n";
} catch {
    print "in the catch, got ($_)\n";
    die "died in catch ($_)";
    print "still printing in the CATCH after death\n";
} finally {
    $var = 1;
    print "in the finally, got ($_)\n";
    die "died in finally ($_)";
    print "still printing in the FINALLY after death\n";
} finally {
    $var = 2;
    print "in the SECOND finally, got ($_)\n";
}
finally {
    $var = 3;
    print "in the THIRD finally, got ($_)\n";
}

finally {
    $var = 4;
    print "in the FOURTH finally, got ($_)\n";
};

print "done with code, var is ($var)\n";`
```

```
in the try
in the catch, got (died in try at test.pl line 12.
)
died in catch (died in try at test.pl line 12.
) at test.pl line 16.
Use of uninitialized value $_ in concatenation (.) or string at test.pl line 34.
in the FOURTH finally, got ()
Use of uninitialized value $_ in concatenation (.) or string at test.pl line 29.
in the THIRD finally, got ()
Use of uninitialized value $_ in concatenation (.) or string at test.pl line 25.
in the SECOND finally, got ()
Use of uninitialized value $_ in concatenation (.) or string at test.pl line 20.
in the finally, got ()
Use of uninitialized value $_ in concatenation (.) or string at test.pl line 21.
        (in cleanup) died in finally () at test.pl line 21.
```

gsanders replied:
The weird part from above, is that multiple `finally` blocks seem to execute in Last In First Out order. So our first `finally` should commit, and our second finally should try to delete everything. I think.

gsanders replied:
I think the ultimate answer here is that we can do the `finally` blocks in either order, as long as they are in two separate blocks. In this way, deleting the allocations and doing the transaction commit will both have a chance to succeed independently of each other.

brummett replied:
Having multiple finally blocks and depending on their execution order seems like a code smell.  The docs for Try::Tiny don't say anything about the order, and includes this
```
You must always do your own error handling in the finally block. "Try::Tiny" will not do anything about handling possible errors coming from code located in these blocks.
```
Which sounds like you can't even count on all the finally blocks running if you throw an exception in one of them.

Using nested try/catch/finally blocks would be safer and might be more clear about what you're trying to do and when.

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -219,24 +226,70 @@ sub create {
+        unless (eval { $self->_disk_allocation->reallocate }) {
+            $self->warning_message("Failed to reallocate my disk allocation with id (%s). Eval returned (%s) ", $self->_disk_allocation->id, $@);
+        }
+        $self->output_dir($self->_disk_allocation->absolute_path); #update if was moved
```
It seems like this little piece of code to reallocate the disk allocation of a software result and update the output_dir if necessary belongs in SoftwareResult.pm  (It seems weird to me for child classes to be modifying output_dir at whim)

gsanders replied:
We totally agree... and I've almost pushed this change. Except I just noticed that SoftwareResult.pm has `disk_allocations`, and both Merged.pm and AlignmentResult.pm have `_disk_allocation`. And all of the above are optional and is_many. So... huh. Inheritance is weird in this area? We are still looking at it.

gsanders replied:
```
02/24/15 14:57:32 [7] my $test = Genome::SoftwareResult->get('9e423b7a7a4d41a8aa581d6d6ba1e816');
$Genome_Disk_Allocation1 = Genome::Disk::Allocation=HASH(0x87b30c8);

02/24/15 14:57:36 [8] $test->_disk_allocation;
$Genome_Disk_Allocation1 = Genome::Disk::Allocation=HASH(0x87b30c8);

02/24/15 14:57:39 [9]  $test->disk_allocations
$Genome_Disk_Allocation1 = Genome::Disk::Allocation=HASH(0x87b30c8);
```

I don't even know what this means or how anything works anymore :)

gsanders replied:
I asked Tom and his recollection is that AlignmentResult was the first ever software result and thus had it's own disk allocation. Later, SR contained its own accessor but the code was never deduplicated. I believe in practice the two values are always the same.

As such I've just chosen to use the SR level accessor, and hope it will always work until we deduplicate the code in another pull request another time. Let us know what you think. It has been pushed.

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -297,6 +368,15 @@ sub collect_individual_alignments {
+        my $test_name;
+        if ($ENV{GENOME_SOFTWARE_RESULT_TEST_NAME}) {
+            $test_name = $ENV{GENOME_SOFTWARE_RESULT_TEST_NAME};
+        } elsif ($self->test_name) {
```
I actually think we always want to use the same test name as the merged result.  In what case would we want to use the environment variable?

apregier replied:
In fact, in get_merged_alignment_results on AlignmentResult, you always use the test_name on the current result.

gsanders replied:
I think Feiyu and I both agree with you. We had the same thought when we first reviewed the existing code, and only included this section in case we were wrong.

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -707,4 +786,54 @@ sub scalar_property_from_underlying_alignment_results {
+            }
+
+            unless ($ENV{UR_DBI_NO_COMMIT}) {
+                my $lock_var = File::Spec->join($ENV{GENOME_LOCK_DIR}, 'genome', __PACKAGE__, 'lock-per-lane-alignment-'.$alignment->id);
```
Just trying to think about your locking problem - are you guaranteed that all of the alignment objects returned by collect_individual_alignments will be unique?

dufeiyu replied:
Resolved by adding "once => 1" to UR::Observer

apregier commented on lib/perl/Genome/InstrumentData/AlignmentResult_regenerate.t:
```diff
@@ -0,0 +1,239 @@
+my $merge_class      = 'Genome::Test::Factory::InstrumentData::MergedAlignmentResult';
+my $allocation_class = 'Genome::Test::Factory::DiskAllocation';
+
+# We need to override this because when an allocation is moved the contents do not seem to be copied. A bug or something we messed up? FIXME 
```
Have you had a chance to come back to this FIXME?  (I don't know the answer, sorry!)

gsanders replied:
I think this is now unnecessary. Tests seem to agree! Thanks so much for the catch!

gsanders commented:
We just pushed a few more commits to address PR comments, including documentation changes, cleanup corrections, using ID instead of refaddr. More changes coming soon.

gsanders commented:
I believe all of the comments thus far have been addressed. Mostly with fixes, sometimes with conversation and more questions. But we are done until the next round of comments :). Thanks to everyone and keep it coming!

gsanders commented:
Tomorrow, Feiyu and I are going to close this pull request and create a new one because rebasing will likely be easier than merging because of the long-running nature of this PR and software-result changes that have been made in the meantime.

Pleas feel free to continue to comment here until we close this PR and link to the next one.

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -219,25 +226,65 @@ sub create {
+    #The transaction commit and temporary allocation deletion need to
+    #be in different finally blocks.
+    finally {
+        unless ($_) {
```
I think the test output from [this comment](https://github.com/genome/genome/pull/407/files#r25272900) suggested that `$_` won't be defined inside the `finally`.

If this needs to only commit if there's not an error, it seems like it belongs inside the `try` block itself.

brummett replied:
Finally blocks get passed the exception in @_ if there was an exception, and no args if there was no exception.

apregier replied:
I agree with Tom - can you just move that inside the try block?

dufeiyu replied:
ok, we will check this again and make the change if necessary.

gsanders replied:
Thanks very much for all the insight here. We've pushed a commit that moves the commit up to the try block. We agree this seems much cleaner!

acoffman commented:
@gsanders I don't know how other people feel about it, but you can actually rebase on your current branch and then force push to this PR and it will work just fine if that seems easier to you. That's what I've been doing on my PRs.

dufeiyu commented:
CI, test this

nnutter commented:
Just ignore `apipe-ci/bats` because this branch predates dependencies on it.

gsanders commented:
@acoffman thanks! That does seem easier to me, so I think we'll proceed with that plan.

gsanders commented:
We have merged master, which required some manual work (and reset history for this PR). The necessary data has been added to the test-db, and everything seems to be passing locally.

We are going to run another real-world build test set to make sure nothing sneaky has broken. However, I think we are mostly ready for +1's and merging here.

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -153,8 +155,10 @@ sub create {
+    my @temp_allocations = ();
 
+    try {
         #TODO In a future version collect relevant alignments from other merged alignment results when available
```
Does this `#TODO` describe what this PR is doing?

gsanders replied:
I don't think so, but I don't know the original intent of the TODO! I believe it is quite old.

tmooney replied:
It looks like that `TODO` is from the original version of this module in 2011!  Now that you mention it, I believe the intention was to see if we could find an existing merged result and add our newly aligned lane to it instead of re-merging all the individual lanes again.  You're right that this isn't quite that... I wonder if that's still a thing we might want to do?

gsanders replied:
Ah! That makes sense. That would be really nice, especially with the way things are now with needing to recreate per-lane bams just to do a new merge. I don't know that we'll be able to get to this immediately but I'll certainly keep it in mind for future alignment work.

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -298,6 +358,8 @@ sub collect_individual_alignments {
             };
         }
 
+        my $test_name = $self->test_name // undef;
```
Is the `// undef` just to make it explicit that `undef` is okay?  (That is to say, `undef` will only be assigned if `$self->test_name` returned undef.)

gsanders replied:
Good point! This seems unnecessary. We can take it out.

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -709,4 +770,54 @@ sub scalar_property_from_underlying_alignment_results {
+
+                # If the build before us successfully created a merged alignment result, we no longer need a lock
+                # If it failed, we will add an observer just as the first build did.
+                if ($alignment->get_merged_alignment_results) {
```
Is it worth trying `$alignment->get_merged_alignment_results` once without locking and only locking [and checking again] if it fails?  (This is similar to how `get_with_lock` and the SR User locking work.)

gsanders replied:
Hmm. Good point. I'd have to think about this for little bit. I think I like your suggestion.

If we went forward with that, we'd still check again and (possibly) immediately unlock, correct? I believe that is the case, in the case of contention around the lock, and 'waking up' to find the merged alignment result is now present. So I believe the only code change would be to add one additional check before locking and skip locking if appropriate?

tmooney replied:
Yep, since it's already doing a `load()` it should be enough to check before locking and leave the rest as-is.  (We could always come back and add that later.)

gsanders replied:
Oh! Check out line 777.  We won't be in this code block `$alignment->get_merged_alignment_results` returns something here. I believe this is sufficient, thoughts?

tmooney replied:
Aye, that'd do it!

gsanders commented:
Pushed a commit which allows real-world builds to succeed again after merging in changes from master (Software Result user changes). 

The last remaining item of which we are aware is that we need to delete the per-lane bam and reallocate with a UR Observer / commit hook instead of doing it in-line.

gsanders commented:
We just pushed a fix for the last remaining code review commentary. Please give us some more feedback, or some +1's!

apregier commented:
+1

gsanders commented:
Random tip for those traversing huge pull requests: Add '?w=1' to the URL to ignore whitespace changes. This often will help a great deal with parsing the real changes.

Final URL for the files page: https://github.com/genome/genome/pull/407/files?w=1

gsanders commented:
I just pushed what I *think* is our final clean-up, including one naming change and one commit with a larger purpose (clean up and backfill more easily on production data once we are live).

gsanders commented:
The latest commit should make tests pass. One more code review needed!

tmooney commented on lib/perl/Genome/SoftwareResult.pm:
```diff
@@ -1190,4 +1190,22 @@ sub expunge {
     return 1;
 }
 
+sub resize_disk_allocation {
```
Ideally we don't need both this and `Genome::SoftwareResult::Stageable::_reallocate_disk_allocation`.

gsanders replied:
Fixed and pushed!

tmooney commented on lib/perl/Genome/SoftwareResult.pm:
```diff
@@ -1190,4 +1190,22 @@ sub expunge {
+        unless (eval { $self->disk_allocations->reallocate(%params) }) {
+            $self->warning_message("Failed to reallocate my disk allocation with id (%s). Eval returned (%s) ", $self->disk_allocations->id, $@);
+        }
+        $self->output_dir($self->disk_allocations->absolute_path); #update if was moved
```
This line is handled by `Genome::Disk::Allocation::_update_owner_for_move` now.

gsanders replied:
We are going to add a call to this inside `Genome::SoftwareResult::Stageable::_reallocate_disk_allocation` unless there are any objections. It just seems cleaner to attempt it every time. Thoughts?

gsanders replied:
Fixed!

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult_regenerate.t:
```diff
@@ -0,0 +1,239 @@
+
+subtest 'get_unarchived_merged_alignment_results' => sub {
+    is_deeply([sort $ar1->get_unarchived_merged_alignment_results], [sort($merged_result, $smaller_merged_result)], 'Got two unarchived merged results for ar1');
+    is_deeply([$ar2->get_unarchived_merged_alignment_results], [$merged_result], 'Got one unarchived merged results for ar2');
```
`results` should be singular.

dufeiyu replied:
fixed

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult_regenerate.t:
```diff
@@ -0,0 +1,239 @@
+        value_id => $ar1_instrument_data_id,
+    );
+
+    # need to manually calculte the lookup hash or get_with_lock will fail on mock objects
```
An `a` got lost from `calculate`.

dufeiyu replied:
fixed

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult_regenerate.t:
```diff
@@ -0,0 +1,239 @@
+    map {$_->test_name($test_name); $_->recalculate_lookup_hash} ($ar1, $ar2, $bad_ar, $merged_result, $smaller_merged_result, $bad_merged_result);
+
+    # Set up allocations
+    class GenomeTest::Object{ };
```
What does `GenomeTest::Object` do here?

dufeiyu replied:
Removed.

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.t:
```diff
@@ -19,7 +18,17 @@ BEGIN {
+my $pkg = 'Genome::InstrumentData::AlignmentResult::Merged';
+use_ok($pkg);
+
+# Override methods for testing. We do this here so we don't worry about UR_DBI_NO_COMMIT in actual code.
```
I don't think we can safely disregard `UR_DBI_NO_COMMIT` in the code.  Do we want the callback to fire and remove the per-lane BAM if it is turned on and the merged BAM we are creating won't be saved?

gsanders replied:
Fixed! Thanks for the perceptive catch!

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.t:
```diff
@@ -31,21 +40,18 @@ my $alignment_result_class_name = "Genome::InstrumentData::AlignmentResult::" .
-my $aligner_label   = $aligner_name.$aligner_version;
-$aligner_label =~ s/\./\_/g;
+my $test_name = 'merged_unit_test';
+my $convert_name = Genome::InstrumentData::AlignmentResult->_resolve_subclass_name_for_aligner_name($aligner_name);
```
Was `$convert_name` planned to be used?

dufeiyu replied:
removed.

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.t:
```diff
@@ -31,21 +40,18 @@ my $alignment_result_class_name = "Genome::InstrumentData::AlignmentResult::" .
+$aligner_label       =~ s/\./\_/g;
 
-my $FAKE_INSTRUMENT_DATA_ID=-123456;
+my $expected_base_dir = $ENV{GENOME_TEST_INPUTS} .'/Genome-InstrumentData-AlignmentResult-Merged';
```
This would be better as `Genome::Utility::Test->data_dir_ok($pkg);`.

dufeiyu replied:
Good point. Changed.

tmooney commented on lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm:
```diff
@@ -709,4 +827,52 @@ sub scalar_property_from_underlying_alignment_results {
 }
 
 
+sub _lock_per_lane_alignment {
```
This might be better named `_lock_per_lane_alignments`.  (There's also an argument for `_lock_individual_alignments` since that's what it collects!)

tmooney commented:
+1 :four_leaf_clover: I'm hopeful that this works!
